frame/contracts/src/lib.rs

```diff
@@ -1054,19 +1043,94 @@ impl Default for Schedule {
 	}
 }
 
+#[doc(hidden)]
+#[cfg_attr(test, derive(Debug))]
+pub struct DynamicWeightData<AccountId, NegativeImbalance> {
+	/// The account id who should receive the refund from the gas leftovers.
+	transactor: AccountId,
+	/// The negative imbalance obtained by withdrawing the value up to the requested gas limit.
+	imbalance: NegativeImbalance,
+}
+
 /// `SignedExtension` that checks if a transaction would exhausts the block gas limit.
 #[derive(Encode, Decode, Clone, Eq, PartialEq)]
 pub struct CheckBlockGasLimit<T: Trait + Send + Sync>(PhantomData<T>);
 
+impl<T: Trait + Send + Sync> CheckBlockGasLimit<T> {
+	/// Perform pre-dispatch checks for the given call and origin.
+	fn perform_pre_dispatch_checks(
+		who: &T::AccountId,
+		call: &<T as Trait>::Call,
+	) -> Result<Option<DynamicWeightData<T::AccountId, NegativeImbalanceOf<T>>>, TransactionValidityError> {
+		let call = match call.is_sub_type() {
+			Some(call) => call,
+			None => return Ok(None),
+		};
+
+		match *call {
+			Call::claim_surcharge(_, _) | Call::update_schedule(_) | Call::put_code(_) => Ok(None),
+			Call::call(_, _, gas_limit, _) | Call::instantiate(_, gas_limit, _, _) => {
+				// Compute how much block weight this transaction can take at its limit, i.e.
+				// if this transaction depleted all provided gas to zero.
+				let gas_weight_limit = Weight::try_from(gas_limit)
+					.map_err(|_| InvalidTransaction::ExhaustsResources)?;
+				let weight_available = <T as system::Trait>::MaximumBlockWeight::get()
```

Comment:
Could be made into a function in system: `get_remaining_block_weight`

Reply:
Added! Named it `remaining_weight` though to be inline with other weight related functions in system